st2actions/st2actions/worker.py

```diff
@@ -14,93 +14,62 @@
-from st2common.util.greenpooldispatch import BufferedDispatcher
 
-LOG = logging.getLogger(__name__)
 
+LOG = logging.getLogger(__name__)
 
-ACTIONRUNNER_WORK_Q = liveaction.get_queue('st2.actionrunner.work',
-                                           routing_key=publishers.CREATE_RK)
+ACTIONRUNNER_WORK_Q = liveaction.get_status_management_queue(
+    'st2.actionrunner.work', routing_key=action_constants.LIVEACTION_STATUS_SCHEDULED)
 
 
-class Worker(ConsumerMixin):
+class ActionExecutionDispatcher(consumers.MessageHandler):
+    message_type = action_models.LiveActionDB
 
-    def __init__(self, connection):
-        self.connection = connection
+    def __init__(self, connection, queues):
+        super(ActionExecutionDispatcher, self).__init__(connection, queues)
         self.container = RunnerContainer()
-        self._dispatcher = BufferedDispatcher()
-
-    def shutdown(self):
-        self._dispatcher.shutdown()
-
-    def get_consumers(self, Consumer, channel):
-        consumer = Consumer(queues=[ACTIONRUNNER_WORK_Q],
-                            accept=['pickle'],
-                            callbacks=[self.process_task])
-        # use prefetch_count=1 for fair dispatch. This way workers that finish an item get the next
-        # task and the work does not get queued behind any single large item.
-        consumer.qos(prefetch_count=1)
-        return [consumer]
-
-    def process_task(self, body, message):
-        # LOG.debug('process_task')
-        # LOG.debug('     body: %s', body)
-        # LOG.debug('     message.properties: %s', message.properties)
-        # LOG.debug('     message.delivery_info: %s', message.delivery_info)
-        try:
-            self._dispatcher.dispatch(self._do_process_task, body)
-        finally:
-            message.ack()
 
-    def _do_process_task(self, body):
-        try:
-            self.execute_action(body)
-        except Exception:
-            LOG.exception('execute_action failed. Message body : %s', body)
-
-    def execute_action(self, liveaction):
-        # Note: We only want to execute actions which haven't completed yet
-        if liveaction.status == LIVEACTION_STATUS_CANCELED:
-            LOG.info('Not executing liveaction %s. User canceled execution.', liveaction.id)
-            if not liveaction.result:
-                update_liveaction_status(status=LIVEACTION_STATUS_CANCELED,
-                                         result={'message': 'Action execution canceled by user.'},
-                                         liveaction_id=liveaction.id)
-            return
+    def process(self, liveaction):
+        if liveaction.status != action_constants.LIVEACTION_STATUS_SCHEDULED:
+            LOG.info('%s is not executing %s (id=%s) with "%s" status.',
+                     self.__class__.__name__, type(liveaction), liveaction.id, liveaction.status)
+
+            if (not liveaction.result and
```

Comment:
Personally, I would find this easier to read and understand if we had 2 "top level" ifs:

1. One which updates result for canceled actions without a result and returns
2. One which returns for all the actions which have already finished